qemu/lib/cdrom.py

```diff
@@ -0,0 +1,162 @@
+"""
+cdrom event check utility functions.
+
+This module is meant to reduce code size by performing extra event check
+procedures. Generally, functions here should provide layers of extra event
+check for cdrom related operations.
+"""
+import logging
+
+from avocado import fail_on
+from virttest import virt_vm, utils_misc
+
+
+def is_device_tray_opened(vm, device_id):
+    """
+    check whether specified block device tray is open or not.
+    Return True, if device is opened, else False.
+
+    : param vm: VM object
+    : device_id: block device identifier
+    """
+    blocks_info = vm.monitor.info('block')
+
+    if isinstance(blocks_info, str):
+        open_str = 'tray open'
+        close_str = 'tray closed'
+        for block in blocks_info.splitlines():
+            if device_id in block:
+                if open_str in block:
+                    return True
+                elif close_str in block:
+                    return False
+    else:
+        for block in blocks_info:
+            if device_id in str(block) and block.get('tray_open'):
+                return block['tray_open']
+    return False
+
+
+class QMPEventCheck(object):
+    """
+    base context manager class.
+    """
+
+    def __init__(self, *args, **kargs):
+        super(QMPEventCheck, self).__init__(*args, **kargs)
+
+    def __enter__(self):
+        """
+        setup of event check.
+        """
+        raise NotImplementedError
+
+    def __exit__(self, *exc_info):
+        """
+        actual event check goes here, after the execution of event-triggering
+        function.
+        """
+        if exc_info[0]:
+            return
+        self._event_check()
+
+    def _event_check(self):
+        raise NotImplementedError
+
+
+class QMPEventCheckCD(QMPEventCheck):
+    """
+    context manager class to handle checking of event "DEVICE_TRAY_MOVED"
+    """
+
+    def __init__(self, vm, device_id, operation):
+        self.vm = vm
+        self.device_id = device_id
+        self.operation = operation
+
+    def __enter__(self):
+        """
+        initialize status of tray and clear events of qmp_monitors.
+        """
+        status = is_device_tray_opened(self.vm, self.device_id)
+        self.status_before = status
+        self.status_after = status
+
+        for m in self.vm.qmp_monitors:
+            m.clear_events()
```

Comment:
For some reason, I'd suggest to only clear the certain type of event instead of all.